### DISCUSSION TITLE
New matching

### DISCUSSION THREAD

**Discussion by swhite2401** [atcollab/at#880] [Python AT]
@lfarv, I start here a discussion to collect remarks on the new matching functions

**Comment by swhite2401** [comment]
First remark: matching mu is really not user friendly, in fact for the moment the only way I managed to do it in the new matching is by creating a user function.

The previous matching had been improve iteratively with several functionalities:
-use_integer: select whether to use or not the integer part, this prevent calculating optics at all refpts and significantly speeds up the calculation
-allow for matching in units in units of 2*pi instead of phase

Second remark: when one defines a refpts that is not found observable returns None and matching proceed, in this case I would raise an AtError: "Refpts not defined for ring" or something similar

**Comment by lfarv** [comment]
@swhite2401 : good idea to group all the requests before defining a new version.

#### phase advance matching
- **integer part**: the `LocalOpticsObservable` has a `all_points` keyword, which indeed significantly changes the computation time. At the moment it is forced to True when asking for `mu`. From the documentation:
  
  _all_points:     Compute the local optics at all elements. This avoids discontinuities in phase advances. This is automatically set for the 'mu' parameter, but may need to be specified for user-defined evaluation functions using the phase advance._

  We can remove the special treatment of `mu`: `all_points` is False by default. This is dangerous: you can never know where the phase advance jumps by 2 $\pi$, but why not.

  Or change the default: `all_points` would default to False, except for the `mu` parameter where it defaults to True. This is safer, the user has to explicitly set `all_points` to False if he accepts phase jumps.

- **units of 2*pi**: add a keyword in `LocalOpticsObservable`? Or create a new parameter name meaning $\frac{\mu}{2\pi}$?
- **user function**: are you matching the phase advance at a given location or the difference of phase advance between 2 points ? Then have a look at [this](https://atcollab.github.io/at/p/notebooks/observables.html), it's quite simple. Though we can look for another solution…

**Reply by swhite2401** [reply]
The most use case is to fit the fractional part, it is very rare that one would change integer.

My proposal would then be to create additional parameters:
-keep mu as is, compute at all refpts and include integer part
-add muf, compute only at needed refpts and match only the fractional part
-mu2pi, and muf2pi would allow matching in units of 2pi (which is what one would naturally do)

all_points would be set to True only if mu or mu2pi are used for matching

I was matching the phase at a given point in units of 2pi, but there may be a bug there either in the matching or in my definition because x and y plane were returning the same value. So in the end I had to write separate x and y functions.

**Comment by lfarv** [comment]
I create a branch improved_matching to implement the modifications discussed here. It's a work in progress but it allows to check if the modifications solve the issues. For the moment:

- The 3 new parameters "muf", "mu2pi", "mu2pif" are implemented and documented in the `LocalOpticsObservable`
- The "empty" observable, when its refpts does not refer to any location is treated in the following way: in `LocalOpticsObservable`, it behaves correctly: the value has 0 lines and the expected number of columns, deviation and residuals are accordingly empty and the sum of residuals gives 0. Since this may lead to an infinite search in matching, a new `check` method of `Observable` and `ObservableList` has been added. Is is called in the initialisation of matching and raises an exception in that case.

**Comment by lfarv** [comment]
@swhite2401 : I found a problem in matching:

Initially, the `ObservableList.evaluate()` method did not return anything. You noticed that it would better return the evaluated observable values. This looked so natural that I forgot the reason it was initially done so. And it turns out that this modification may break the matching:

During the evaluation of a list of observables, it may happen that an exception prevents from getting a value: unstable lattice for instance. When this happens, the observable is flagged as failing by storing the exception instead of the result, and the evaluation goes on for the remaining observables. Finally, `evaluate()` terminates normally.

Then when asking for the `value` property of each observable, the stored exception, if any, is raised. The `deviation` attribute may either raise also the exception or return a default (high) value if one is provided. This is how the matching can survive an "unstable machine" exception while trying to get away from the "forbidden" area:
```python
        constraints.evaluate(ring, dp=dp, dct=dct, df=df)
        return constraints.get_flat_weighted_deviations(err=1.0e6)
```

But now, the exception is raised while `evaluate()` returns the values, which stops the matching immediately.

There are several possibilities to correct that:

1. Go back to the initial state: `evaluate()` does not return anything. Simple and clean, but unfortunately the modification is already in the master branch and this would break the compatibility,
2. Catch the evaluate() exception in matching and continue. This is rather inelegant: why throwing exceptions and ignore them?
3. Add a new "safe_evaluate" method identical to the initial `evaluate()` and use it in matching,
4. Return "non-breaking" values from `evaluate()`. Such values contain an error code instead of the value when the evaluation failed. It's how the values are internally stored. I don't like that: it requires the user to check the value before using it, it's like moving back from modern exception handling to old-style error codes,
5. instead of returning the list of observable values, evaluate() would return a list-like object "`result`" such that the exception would be raised (if relevant) at the time each item `result[i]` is accessed. It's like a clean variant of solution 4. It ensures the compatibility but needs some more work. This is my preferred solution.

Whatever the solution, it will need to be back-ported to the master branch before the "improved_matching" branch is merged.

**Comment by lfarv** [comment]
#892 is created to implement step by step the conclusions of this discussion.
The issue mentioned above is solved according to solution 5 so that the 'improved_matching" branch is fully compatible with the "master: branch.

Up to now, all remarks are taken into account, you are welcome to test the branch when possible.
